langchain/src/embeddings/googlevertexai-image.ts

```diff
@@ -0,0 +1,143 @@
+import { Embeddings, EmbeddingsParams } from "./base.js";
+import {
+  GoogleVertexAIBasePrediction,
+  GoogleVertexAIConnectionParams,
+  GoogleVertexAILLMResponse,
+} from "../types/googlevertexai-types.js";
+import { GoogleVertexAIConnection } from "../util/googlevertexai-connection.js";
+import { AsyncCallerCallOptions } from "../util/async_caller.js";
+// import { chunkArray } from "../util/chunk.js";
+
+export interface GoogleVertexAIImageEmbeddingsParams
+  extends EmbeddingsParams,
+    GoogleVertexAIConnectionParams {}
+
+interface GoogleVertexAIImageEmbeddingsOptions extends AsyncCallerCallOptions {}
+
+interface GoogleVertexAIImageEmbeddingsInstance {
+  text?: string;
+  image?: {
+    bytesBase64Encoded: string;
+  };
+}
+
+interface GoogleVertexImageEmbeddingsResults
+  extends GoogleVertexAIBasePrediction {
+  textEmbedding?: number[];
+  imageEmbedding?: number[];
+}
+
+/**
+ * The media should have a text property, an image property, or both.
+ */
+export type GoogleVertexAIMedia =
+  | {
+      text: string;
+      image?: Buffer;
+    }
+  | {
+      text?: string;
+      image: Buffer;
+    };
+
+export type MediaEmbeddings = {
+  text?: number[];
+  image?: number[];
+};
+
+export class GoogleVertexAIImageEmbeddings
+  extends Embeddings
+  implements GoogleVertexAIImageEmbeddingsParams
+{
+  model = "multimodalembedding@001";
+
+  private connection: GoogleVertexAIConnection<
+    GoogleVertexAIImageEmbeddingsOptions,
+    GoogleVertexAIImageEmbeddingsInstance,
+    GoogleVertexImageEmbeddingsResults
+  >;
+
+  constructor(fields?: GoogleVertexAIImageEmbeddingsParams) {
+    super(fields ?? {});
+
+    this.model = fields?.model ?? this.model;
+
+    this.connection = new GoogleVertexAIConnection(
+      { ...fields, ...this },
+      this.caller
+    );
+  }
+
+  mediaToInstance(
+    media: GoogleVertexAIMedia
+  ): GoogleVertexAIImageEmbeddingsInstance {
+    const ret: GoogleVertexAIImageEmbeddingsInstance = {};
+
+    if (media?.text) {
+      ret.text = media.text;
+    }
+
+    if (media.image) {
+      ret.image = {
+        bytesBase64Encoded: media.image.toString("base64"),
+      };
+    }
+
+    return ret;
+  }
+
+  responseToEmbeddings(
+    response: GoogleVertexAILLMResponse<GoogleVertexImageEmbeddingsResults>
+  ): MediaEmbeddings[] {
+    return response.data.predictions.map((r) => ({
+      text: r.textEmbedding,
+      image: r.imageEmbedding,
+    }));
+  }
+
+  async embedMedia(media: GoogleVertexAIMedia[]): Promise<MediaEmbeddings[]> {
+    // Only one media embedding request is allowed
+    return Promise.all(media.map((m) => this.embedMediaQuery(m)));
+  }
+
+  async embedMediaQuery(media: GoogleVertexAIMedia): Promise<MediaEmbeddings> {
```

Comment:
Is there a purpose to embedding both at the same time? Does one input influence the result of the other?

Reply:
The API allows for it is the biggest reason. The two vectors are independent of each other, except they live in the same vector space.

My guess is that the original purpose was so you could submit an image and a text, get the resultant vectors, and compute the similarity between the two all in one shot - which should be faster than two queries. I saw at least one other model that did this as well.

Reply:
I think I'd rather keep them as separate methods and add the combined one in if there's some benefit to reduce confusion for now - this one is already going to be an odd duck.